libipl/p10/ipl0.C

```diff
@@ -374,28 +387,137 @@ static int ipl_alignment_check(void)
+				pdbg_target_path(clock_target), status);
+
+			ipl_error_callback(IPL_ERR_TGT_PROBE_FAILED);
+			rc++;
+			continue;
+		}
+
+		// Resetting the clock chip, so that it will recalibrate input oscillator
+		// signal and identifies bad oscillators.
+		data = OSC_RESET_CMD;
+		i2c_rc = i2c_write(clock_target, 0, OSC_CTL_OFFSET, OSC_RESET_CMD_LEN, &data);
+		if(i2c_rc){
+			ipl_log(IPL_ERROR, "soft reset command is failed for clock '%s' with rc = %d\n",
+				pdbg_target_path(clock_target), rc);
+
+			ipl_error_callback(IPL_ERR_CLK_FAILED);
+			rc++;
+			continue;
+		}
+		else{
+			ipl_log(IPL_DEBUG, "soft reset command is successfull for clock '%s'\n",
+				pdbg_target_path(clock_target));
+		}
+
+		// wait for clock to restart
+		std::this_thread::sleep_for(std::chrono::milliseconds(CLOCK_RESTART_DELAY_IN_MS));
+
+		// Read clock status register to check whether it reports calibration error.
+		// Bit-0 will be set if there is a calibration error.
+		i2c_rc = i2c_read(clock_target, 0, OSC_STAT_OFFSET, OSC_STAT_REG_LEN, &data);
+		if(i2c_rc){
+			ipl_log(IPL_ERROR, "status register read is failed for clock '%s', with rc = %d\n",
+				pdbg_target_path(clock_target), rc);
+
+			ipl_error_callback(IPL_ERR_CLK_FAILED);
+			rc++;
+			continue;
+		}
+		else{
+			ipl_log(IPL_DEBUG, "status register value for clock '%s' is 0x%2X\n",
+				pdbg_target_path(clock_target), data);
+
+			if(data & OSC_STAT_ERR_MASK){
+				ipl_log(IPL_ERROR, "Calibration is failed for clock '%s', status=0x%2X",
+					pdbg_target_path(clock_target), data);
+
+				ipl_error_callback(IPL_ERR_CLK_FAILED);
+				rc++;
+				continue;
+			}
+		}
+	}
+	return rc;
+}
+
 static int ipl_set_ref_clock(void)
 {
 	struct pdbg_target *proc;
 	int rc = 0;
+	uint8_t clock_count = 0;
 
 	if (ipl_type() == IPL_TYPE_MPIPL)
 		return -1;
 
 	ipl_log(IPL_INFO, "Istep: set_ref_clock: started\n");
 
+	if(initialize_and_check_clock_chip(clock_count)){
+		ipl_log(IPL_ERROR, "Clock initialization failed\n");
+		return 1;
+	}
+
+	if(clock_count != 1 && clock_count != 2){
+		ipl_log(IPL_ERROR, "Invalid number (%d) of clock target found\n",
+							clock_count);
+
+		ipl_error_callback(IPL_ERR_CLK_FAILED);
+		return 1;
+	}
+
 	pdbg_for_each_class_target("proc", proc) {
 		fapi2::ReturnCode fapirc;
 
 		if (!ipl_is_functional(proc))
 			continue;
 
+		// Check if system have clocks to enable redundant mode,
+		// If yes set attribute to enable redundant mode.
+		// Default value of attribute will be for non-redundant mode
+		if(clock_count == NUM_CLOCK_FOR_REDUNDANT_MODE){
+			fapi2::ATTR_CP_REFCLOCK_SELECT_Type clock_select =
+				fapi2::ENUM_ATTR_CP_REFCLOCK_SELECT_BOTH_OSC0;
+			if(!pdbg_target_set_attribute(
+					proc, "ATTR_CP_REFCLOCK_SELECT", 1, 1, &clock_select)){
+
+				ipl_log(IPL_ERROR, "Attribute CP_REFCLOCK_SELECT update failed"
+					" for proc %d \n", pdbg_target_index(proc));
+				ipl_error_callback(IPL_ERR_ATTR_READ_FAIL);
```

Comment:
Looks, you are trying to write the attribute, right? But, returning the error code as `IPL_ERR_ATTR_READ_FAIL`.

You should use `IPL_ERR_ATTR_WRITE_FAIL` (if not exist, please add it).

Reply:
do we really need seperate return code?
can i change IPL_ERR_ATTR_READ_FAIL to IPL_ERR_ATTR_ACCESS_FAILED?

Reply:
Yeah, that is also fine, I don't think we will treat read/write failure separately, and even if needs we can change it later.

Reply:
Ok